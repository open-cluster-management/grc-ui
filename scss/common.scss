/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/


@import 'node_modules/carbon-components/scss/globals/scss/styles';
@import 'node_modules/carbon-components/scss/globals/scss/_colors';
@import 'node_modules/carbon-components/scss/globals/scss/_helper-mixins';
@import 'header-carbon-v9-migration';
@import 'variables';
@import '_red-hat-font';

@include redhat-font-face;

html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video {
  font-family: RedHatText;
}

.bx--label, .bx--search-input, .bx--overflow-menu-options__btn, .bx--pagination__text, .bx--text-input, .bx--tooltip--icon__top::before, .bx--tooltip--icon__top::after, .bx--tooltip--icon__bottom::before, .bx--tooltip--icon__bottom::after {
  font-family: RedHatText;
}

html {
  height: 100%;
}

body {
  min-height: 100%;
  height: 100%;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  .bx--data-table-v2--zebra tbody tr:nth-child(2n) {
    background-color: $color__white;
  }
  .bx--tooltip--shown {
    font-size: .875rem;
  }
}

.page-content-container {
  display: flex;
  flex-direction: column;
  padding: 1.5rem 5% 3.375rem;

  .bx--select--inline{
    svg.bx--select__arrow{
      position: absolute
    }
  }
}

.content-spinner {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.resource-list-filter {
  margin-top: -1.5rem;
  padding-bottom: 3rem;
}

.bx--loading-overlay {
  z-index: 10000; // need higher z-index with modals open
}

.expand-vertically {
  height: 100%;
}

.nested-expandable-row {
  display: flex;
  flex-direction: column;
  .nested-expandable-row-header {
    font-weight: bold;
    color: $color__blue-50;
  }
  .nested-expandable-row-content {
    margin-top: -2rem;
  }
}

.no-resource {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 32.375rem;
  height: 25.6875rem;
  color: $color__blue-90;
  padding: 1.25rem 2.5rem;
  background-color: transparent;
  box-shadow: none;
  border: none;
  margin: 3rem auto 1rem auto;

  .button-group {
    display: flex;
    .buttons {
      margin: 0 1rem;
      button:hover {
        .doc-link {
          color: white;
        }
      }
      .doc-link {
        text-decoration: none;
      }
    }
  }

  .no-resource-icon {
    width: 22.5rem;
    height: 12.5rem;
  }

  .no-resource-title {
    font-weight: 500;
    margin-bottom: 0.5625rem;
    font-size: 1.8125rem;
    text-align: center;
    color: $grc_color_blue;
  }

  .no-resource-detail {
    font-weight: 300;
    font-size: 1rem;
    margin-bottom: 1.3125rem;
    text-align: center;
    line-height: 1.2;
    color: $color__navy-gray-6;

    > a {
      font-weight: 700;
      margin-right: 0.3125rem;
      font-size: .875rem;
    }
  }
}

.yamlEditorContainer {
  #brace-editor {
    min-width: 31.25rem;
    min-height: 31.25rem
  }
}
