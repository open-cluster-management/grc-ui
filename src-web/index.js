/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { BrowserRouter } from 'react-router-dom'
import { ApolloProvider } from 'react-apollo'
import { App } from './containers/App'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import GrcApolloClient from '../lib/client/apollo-client'
import ScrollToTop from './components/common/ScrollToTop'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

const loggerMiddleware = createLogger()
// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (!window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ && config['featureFlags:reduxLogger']) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__

// Create Redux store with initial state
const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
  applyMiddleware(...middleware)
))



render(
  <ApolloProvider client={GrcApolloClient.getGrcClient()}>
    <Provider store={store}>
      <BrowserRouter>
        <ScrollToTop>
          <App />
        </ScrollToTop>
      </BrowserRouter>
    </Provider>
  </ApolloProvider>
  , document.getElementById('page')
)
