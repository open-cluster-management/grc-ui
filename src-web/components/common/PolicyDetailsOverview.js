/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { Notification, Loading } from 'carbon-components-react'
import { connect } from 'react-redux'
import { updateResourceToolbar } from '../../actions/common'
import msgs from '../../../nls/platform.properties'
import resources from '../../../lib/shared/resources'
import _ from 'lodash'
import DetailsModule from './DetailsModule'
import SimpleTable from './SimpleTable'
// eslint-disable-next-line import/no-named-as-default
import ResourceTableModule from './ResourceTableModuleFromProps'
import StructuredListModule from '../../components/common/StructuredListModuleWithActions'

resources(() => {
  require('../../../scss/policy-details-overview.scss')
})

export class PolicyDetailsOverview extends React.PureComponent{
  constructor(props) {
    super(props)
  }

  static propTypes = {
    error: PropTypes.any,
    item: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.object,
      PropTypes.array
    ]),
    loading: PropTypes.any,
    location: PropTypes.object,
    // params: PropTypes.object,
    refreshControl: PropTypes.object,
    resourceType: PropTypes.object,
    staticResourceData: PropTypes.object,
    updateResourceToolbar: PropTypes.func,
  }

  static contextTypes = {
    locale: PropTypes.string
  }

  getTemplates = (spec = {}, templateType = '') => {
    const templates = []
    Object.entries(spec || []).forEach(([key, value]) => {
      if (key.endsWith(`${templateType}-templates`)) {
        value.forEach((item) => {
          templates.push({ ...item.objectDefinition, templateType: key })
        })
      }
    })
    return templates
  }

  getClusterStatus = (item) => {
    const rawStatus = _.get(item, 'raw.status.status', '-')
    const status = {}
    _.forEach(rawStatus, (value) => {
      const clusterNamespace = _.get(value, 'clusternamespace')
      const compliant = _.get(value, 'compliant', 'unknown')
      status[clusterNamespace] = compliant
    })
    return status
  }

  componentWillReceiveProps(nextProps) {
    const {refreshControl, item, updateResourceToolbar:localUpdateResourceToolbar} = nextProps
    if (!_.isEqual(refreshControl, this.props.refreshControl) ||
        !_.isEqual(item, this.props.item)) {
      localUpdateResourceToolbar(refreshControl, {})
    }
  }

  render() {
    const {staticResourceData, resourceType, error, loading, location} = this.props
    let { item:localItem } = this.props
    const { locale } = this.context

    if(error) {
      return <Notification
        title=''
        className='persistent'
        subtitle={msgs.get(error, locale)}
        kind='error' />
    } else if (loading || !localItem || !Array.isArray(localItem)) {
      return <Loading withOverlay={false} className='content-spinner' />
    } else {
      localItem = localItem[0]
    }

    const clusterStatus = this.getClusterStatus(localItem)

    const modulesSecond = [
      <StructuredListModule
        key='placementPolicies'
        rowsKey='placementPoliciesKeys'
        actions={['table.actions.edit']}
        staticResourceData={staticResourceData}
        clusterStatus={clusterStatus}
        location={location}
        resourceType={resourceType}
        textEditButton
        left
      />,
      <StructuredListModule
        key='placementBindings'
        rowsKey='placementBindingKeys'
        actions={['table.actions.edit']}
        staticResourceData={staticResourceData}
        resourceType={resourceType}
        textEditButton
        right
      />,
    ]
    const modulesBottom = []
    const templates = (localItem && localItem.raw && localItem.raw.spec) ? this.getTemplates(localItem.raw.spec) : []
    let templateType

    if (templates){
      const roleTemplates = []
      const policyTemplates = []
      const objectTemplates = []
      for( const template of templates){
        templateType = template.templateType
        switch( templateType ){
        case 'role-templates':
          roleTemplates.push(template)
          break
        case 'policy-templates':
          policyTemplates.push(template)
          break
        case 'object-templates':
          objectTemplates.push(template)
          break
        default:
          break
        }
      }

      const tableResource = {}
      tableResource['role-templates'] = roleTemplates
      tableResource['policy-templates'] = policyTemplates
      tableResource['object-templates'] = objectTemplates
      if(!_.isEmpty(roleTemplates)){
        _.forEach(roleTemplates,(item) => {
          item.id = item.metadata.name
          const subItems = []
          _.forEach(item.rules, (rule) => {
            subItems.push({
              id: _.get(rule, 'complianceType','-'),
              cells: [
                JSON.stringify(_.get(rule, 'complianceType','-')).replace(/\[|\]|"/g, ''),
                JSON.stringify(_.get(rule, 'policyRule.apiGroups','-')).replace(/\[|\]|"/g, ''),
                JSON.stringify(_.get(rule, 'policyRule.resources','-')).replace(/\[|\]|"/g, ''),
                JSON.stringify(_.get(rule, 'policyRule.verbs','-')).replace(/\[|\]|"/g, ''),
              ]
            })
          })
          item.subItems = subItems
        })
        modulesBottom.push(
          <div className='vertical-expend overview-content-bottom' key='role-temp-container'>
            <h5 className='section-title'>{msgs.get('table.header.roleTemplates', locale)}</h5>
            <div className='overview-content-bottom'>
              {React.createElement(SimpleTable,
                {
                  headerRows: ['', ..._.map(staticResourceData.roleTemplates.tableKeys, 'msgKey')],
                  data: roleTemplates,
                  rows: staticResourceData.roleTemplates.rows,
                  subHeaders: staticResourceData.roleTemplates.subHeaders,
                  listSubItems: true,
                  emptyColFront: 1
                })}
            </div>
          </div>
        )
      }
      if(!_.isEmpty(policyTemplates)){
        modulesBottom.push(
          <div className='vertical-expend overview-content-bottom' key='policy-temp-container'>
            <h5 className='section-title'>{msgs.get('table.header.policyTemplates', locale)}</h5>
            <div className='overview-content-bottom'>
              {React.createElement(ResourceTableModule,
                {
                  key: templateType,
                  definitionsKey: 'policyTemplates',
                  staticResourceData: staticResourceData,
                  resourceType: resourceType,
                  resourceData: tableResource,
                  showModuleHeader: false,
                  showSearch: false,
                  showPagination:false
                })}
            </div>
          </div>
        )
      }
      if(!_.isEmpty(objectTemplates)){
        modulesBottom.push(
          <div className='vertical-expend overview-content-bottom' key='obj-temp-container'>
            <h5 className='section-title'>{msgs.get('table.header.objectTemplates', locale)}</h5>
            <div className='overview-content-bottom'>
              {React.createElement(ResourceTableModule,
                {
                  key: templateType,
                  definitionsKey: 'objectTemplates',
                  staticResourceData: staticResourceData,
                  resourceType: resourceType,
                  resourceData: tableResource,
                  showModuleHeader: false,
                  showSearch: false,
                  showPagination:false
                })}
            </div>
          </div>
        )
      }
    }

    let itemPP = '-', itemPB = '-'
    if (localItem && localItem.placementPolicies && Array.isArray(localItem.placementPolicies)) {
      itemPP = localItem.placementPolicies[0]
    }
    if (localItem && localItem.placementBindings && Array.isArray(localItem.placementBindings)) {
      itemPB = localItem.placementBindings[0]
    }
    return (
      <div className='overview-content'>
        <div className='vertical-expend'>
          <h5 className='section-title'>{msgs.get('table.header.policyDetails', locale)}</h5>
          <DetailsModule
            numRows = {3}
            numColumns = {3}
            listData = {localItem}
            listItem = {staticResourceData.detailKeys.rows}
            title = {staticResourceData.detailKeys.title}
            showHeader={false}
          />
        </div>

        <div className='vertical-expend'>
          <h5 className='section-title'>{msgs.get('table.header.placement', locale)}</h5>
          {modulesSecond.length > 0 &&
          <div className='overview-content-second'>
            <div className='overview-content-second-left'>
              {React.cloneElement(modulesSecond[0], { ...staticResourceData.placementPolicyKeys.detailKeys, data:itemPP })}
            </div>
            <div className='overview-content-second-left'>
              {React.cloneElement(modulesSecond[1], { ...staticResourceData.placementBindingKeys.detailKeys, data:itemPB })}
            </div>
          </div>}
        </div>

        {modulesBottom}

      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const {resourceToolbar: {activeFilters}} = state
  return { activeFilters }
}

const mapDispatchToProps = dispatch => {
  return {
    updateResourceToolbar: (refreshControl) => dispatch(updateResourceToolbar(refreshControl, {})),
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PolicyDetailsOverview))
