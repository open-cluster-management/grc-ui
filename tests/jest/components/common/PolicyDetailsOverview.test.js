/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import { PolicyDetailsOverview } from '../../../../src-web/components/common/PolicyDetailsOverview'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import * as reducers from '../../../../src-web/reducers'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import thunkMiddleware from 'redux-thunk'
import apolloClient from '../../../../lib/client/apollo-client'
import { ApolloProvider } from 'react-apollo'
import { Provider } from 'react-redux'
import { staticResourceDataPolicyOverview, itemPolicyOverview } from './CommonTestingData'

describe('PolicyDetailsOverview component', () => {
  it('renders as expected', () => {
    const preloadedState = window.__PRELOADED_STATE__
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const location = {
      'pathname': '/multicloud/policies/policy/cluster1/1569249226915-policy-test',
      'search': '',
      'hash': '',
      'key': 'q1uagn'
    }
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const refreshControl = {
      'reloading': false,
      'refreshCookie': 'grc-refresh-interval-cookie',
      'timestamp': 'Tue Sep 24 2019 09:56:26 GMT-0400 (Eastern Daylight Time)'
    }
    const component = renderer.create(
      <ApolloProvider client={apolloClient.getGrcClient()}>
        <Provider store={store}>
          <BrowserRouter>
            <PolicyDetailsOverview
              item={itemPolicyOverview}
              updateResourceToolbar={jest.fn()}
              staticResourceData={staticResourceDataPolicyOverview}
              location={location}
              resourceType={resourceType}
              refreshControl={refreshControl}
              error={null}
              loading={false}
            />
          </BrowserRouter>
        </Provider>
      </ApolloProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('PolicyDetailsOverview component', () => {
  it('renders as expected', () => {
    const preloadedState = window.__PRELOADED_STATE__
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const location = {
      'pathname': '/multicloud/policies/policy/cluster1/1569249226915-policy-test',
      'search': '',
      'hash': '',
      'key': 'q1uagn'
    }
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const refreshControl = {
      'reloading': false,
      'refreshCookie': 'grc-refresh-interval-cookie',
      'timestamp': 'Tue Sep 24 2019 09:56:26 GMT-0400 (Eastern Daylight Time)'
    }
    const component = renderer.create(
      <ApolloProvider client={apolloClient.getGrcClient()}>
        <Provider store={store}>
          <BrowserRouter>
            <PolicyDetailsOverview
              item={null}
              updateResourceToolbar={jest.fn()}
              staticResourceData={staticResourceDataPolicyOverview}
              location={location}
              resourceType={resourceType}
              refreshControl={refreshControl}
              error={null}
              loading={false}
            />
          </BrowserRouter>
        </Provider>
      </ApolloProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('PolicyDetailsOverview component', () => {
  it('renders as expected', () => {
    const preloadedState = window.__PRELOADED_STATE__
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const location = {
      'pathname': '/multicloud/policies/policy/cluster1/1569249226915-policy-test',
      'search': '',
      'hash': '',
      'key': 'q1uagn'
    }
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const refreshControl = {
      'reloading': false,
      'refreshCookie': 'grc-refresh-interval-cookie',
      'timestamp': 'Tue Sep 24 2019 09:56:26 GMT-0400 (Eastern Daylight Time)'
    }
    const component = renderer.create(
      <ApolloProvider client={apolloClient.getGrcClient()}>
        <Provider store={store}>
          <BrowserRouter>
            <PolicyDetailsOverview
              item={['','','','']}
              updateResourceToolbar={jest.fn()}
              staticResourceData={staticResourceDataPolicyOverview}
              location={location}
              resourceType={resourceType}
              refreshControl={refreshControl}
              error={null}
              loading={false}
            />
          </BrowserRouter>
        </Provider>
      </ApolloProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

