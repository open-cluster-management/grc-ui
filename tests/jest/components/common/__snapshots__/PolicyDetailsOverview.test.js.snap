// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyDetailsOverview component renders as normal 1`] = `
<div
  className="overview-content"
>
  <div
    className="vertical-expend"
  >
    <dl
      aria-label=""
      className="pf-c-accordion"
    >
      <dt>
        <button
          aria-expanded={true}
          className="pf-c-accordion__toggle pf-m-expanded header-container"
          id="compliance.details-toggle"
          onClick={[Function]}
        >
          <span
            className="pf-c-accordion__toggle-text"
          >
            Policy details
          </span>
          <span
            className="pf-c-accordion__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
              />
            </svg>
          </span>
        </button>
      </dt>
      <dd
        aria-label=""
        className="pf-c-accordion__expanded-content pf-m-expanded"
        hidden={false}
        id="compliance.details-expand"
      >
        <div
          className="pf-c-accordion__expanded-content-body"
        >
          <dl
            className="pf-c-description-list pf-m-horizontal pf-m-1-col"
          >
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Name
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Namespace
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                />
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Enforcement
                  <svg
                    aria-describedby="pf-tooltip-1"
                    className="info-icon"
                  >
                    <use
                      href="#diagramIcons_info"
                    />
                  </svg>
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
          </dl>
          <hr
            className="pf-c-divider pf-m-vertical"
          />
          <dl
            className="pf-c-description-list pf-m-horizontal pf-m-1-col"
          >
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Exclude namespaces
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Include namespaces
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Cluster violations
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  <div
                    className="violationCell"
                  >
                    <svg
                      aria-describedby="pf-tooltip-2"
                      aria-hidden={true}
                      aria-labelledby={null}
                      fill="#f0ab00"
                      height="1em"
                      role="img"
                      style={
                        Object {
                          "verticalAlign": "-0.125em",
                        }
                      }
                      viewBox="0 0 576 512"
                      width="1em"
                    >
                      <path
                        d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      />
                    </svg>
                    -
                  </div>
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Categories
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Controls
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
            <div
              className="pf-c-description-list__group"
            >
              <dt
                className="pf-c-description-list__term"
              >
                <span
                  className="pf-c-description-list__text"
                >
                  Standards
                </span>
              </dt>
              <dd
                className="pf-c-description-list__description"
              >
                <div
                  className="pf-c-description-list__text"
                >
                  -
                </div>
              </dd>
            </div>
          </dl>
        </div>
      </dd>
    </dl>
  </div>
  <div
    className="vertical-expend"
  >
    <h5
      className="section-title"
    >
      Placement
    </h5>
    <div
      className="overview-content-second"
    >
      <div
        className="overview-content-second-cell"
      >
        <div
          className="bx--module bx--module--double structured-list-module"
        >
          <div
            className="bx--module__inner"
          >
            <div
              className="bx--module__header"
              style={
                Object {
                  "justifyContent": "space-between",
                }
              }
            >
              <h1
                className="bx--module__title"
              >
                Placement rule
              </h1>
              <div
                aria-expanded={false}
                aria-haspopup={true}
                aria-label="list of options"
                className="bx--overflow-menu"
                onBlur={[Function]}
                onClick={[Function]}
                onClose={[Function]}
                onKeyDown={[Function]}
                open={false}
                role="menu"
                tabIndex={0}
              >
                <svg
                  alt="Actions"
                  aria-label="Actions"
                  className="bx--overflow-menu__icon"
                  fillRule="evenodd"
                  focusable="false"
                  height="15"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="img"
                  viewBox="0 0 3 15"
                  width="3"
                >
                  <title>
                    Actions
                  </title>
                  <path
                    d="M0 1.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0M0 7.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0M0 13.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0"
                  />
                </svg>
              </div>
            </div>
            <div
              className="bx--module__content"
            >
              <section
                aria-label="Placement rule"
                className="bx--structured-list bx--structured-list--condensed"
              >
                <div
                  className="bx--structured-list-tbody"
                  onKeyDown={[Function]}
                >
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Name
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Namespace
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Cluster selector
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Decisions
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      <div
                        className="cluster-status-container"
                      />
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Timestamp
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                </div>
              </section>
            </div>
          </div>
        </div>
      </div>
      <div
        className="overview-content-second-cell"
      >
        <div
          className="bx--module bx--module--double structured-list-module"
        >
          <div
            className="bx--module__inner"
          >
            <div
              className="bx--module__header"
              style={
                Object {
                  "justifyContent": "space-between",
                }
              }
            >
              <h1
                className="bx--module__title"
              >
                Placement binding
              </h1>
              <div
                aria-expanded={false}
                aria-haspopup={true}
                aria-label="list of options"
                className="bx--overflow-menu"
                onBlur={[Function]}
                onClick={[Function]}
                onClose={[Function]}
                onKeyDown={[Function]}
                open={false}
                role="menu"
                tabIndex={0}
              >
                <svg
                  alt="Actions"
                  aria-label="Actions"
                  className="bx--overflow-menu__icon"
                  fillRule="evenodd"
                  focusable="false"
                  height="15"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="img"
                  viewBox="0 0 3 15"
                  width="3"
                >
                  <title>
                    Actions
                  </title>
                  <path
                    d="M0 1.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0M0 7.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0M0 13.5a1.5 1.5 0 1 1 3 0 1.5 1.5 0 1 1-3 0"
                  />
                </svg>
              </div>
            </div>
            <div
              className="bx--module__content"
            >
              <section
                aria-label="Placement binding"
                className="bx--structured-list bx--structured-list--condensed"
              >
                <div
                  className="bx--structured-list-tbody"
                  onKeyDown={[Function]}
                >
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Name
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Namespace
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Placement rule
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Subjects
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                  <div
                    className="bx--structured-list-row"
                  >
                    <div
                      className="bx--structured-list-td"
                    >
                      <p>
                        Timestamp
                      </p>
                    </div>
                    <div
                      className="bx--structured-list-td"
                    >
                      -
                    </div>
                  </div>
                </div>
              </section>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
